# Problem (1) Data Automobile - Regression

### 1. Examine problem 

Menentukan prediksi harga mobil.

### 2. Prepare data

#### Load data automobile

In [1]:
import pandas as pd

df = pd.read_csv("imports-85.data", header=None)
df.head(10)

0    1            2    3      4     5            6    7      8      9   \
0   3    ?  alfa-romero  gas    std   two  convertible  rwd  front   88.6   
1   3    ?  alfa-romero  gas    std   two  convertible  rwd  front   88.6   
2   1    ?  alfa-romero  gas    std   two    hatchback  rwd  front   94.5   
3   2  164         audi  gas    std  four        sedan  fwd  front   99.8   
4   2  164         audi  gas    std  four        sedan  4wd  front   99.4   
5   2    ?         audi  gas    std   two        sedan  fwd  front   99.8   
6   1  158         audi  gas    std  four        sedan  fwd  front  105.8   
7   1    ?         audi  gas    std  four        wagon  fwd  front  105.8   
8   1  158         audi  gas  turbo  four        sedan  fwd  front  105.8   
9   0    ?         audi  gas  turbo   two    hatchback  4wd  front   99.5   

   ...   16    17    18    19    20   21    22  23  24     25  
0  ...  130  mpfi  3.47  2.68   9.0  111  5000  21  27  13495  
1  ...  130  mpfi  3.47  2.68   9.0  111  5000  21  27  16500  
2  ...  152  mpfi  2.68  3.47   9.0  154  5000  19  26  16500  
3  ...  109  mpfi  3.19  3.40  10.0  102  5500  24  30  13950  
4  ...  136  mpfi  3.19  3.40   8.0  115  5500  18  22  17450  
5  ...  136  mpfi  3.19  3.40   8.5  110  5500  19  25  15250  
6  ...  136  mpfi  3.19  3.40   8.5  110  5500  19  25  17710  
7  ...  136  mpfi  3.19  3.40   8.5  110  5500  19  25  18920  
8  ...  131  mpfi  3.13  3.40   8.3  140  5500  17  20  23875  
9  ...  131  mpfi  3.13  3.40   7.0  160  5500  16  22      ?  

[10 rows x 26 columns]

In [2]:
df.shape

(205, 26)

In [3]:
for i in df:
    nr = df[df[i]=='?'].shape[0]
    if nr > 0:
        print(i,' : ', nr)

1  :  41
5  :  2
18  :  4
19  :  4
21  :  2
22  :  2
25  :  4


C:\Users\levono\Anaconda3\lib\site-packages\pandas\core\ops.py:1649: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
  result = method(y)


In [4]:
df.dtypes

0       int64
1      object
2      object
3      object
4      object
5      object
6      object
7      object
8      object
9     float64
10    float64
11    float64
12    float64
13      int64
14     object
15     object
16      int64
17     object
18     object
19     object
20    float64
21     object
22     object
23      int64
24      int64
25     object
dtype: object

In [5]:
import numpy as np

for i in range(len(df.columns)):
    if (i!=5) and (df[i].dtypes=='object') and (df[df[i]=='?'].shape[0]>0):
    #not including att-5 (num-of-doors)
            print('\nAttribute-',i,':',df[df[i]=='?'].shape[0])
            df[i] = df[i].replace('?', np.NaN)
            df[i] = df[i].astype('float64')
                                                
df.dtypes


Attribute- 1 : 41

Attribute- 18 : 4

Attribute- 19 : 4

Attribute- 21 : 2

Attribute- 22 : 2

Attribute- 25 : 4


0       int64
1     float64
2      object
3      object
4      object
5      object
6      object
7      object
8      object
9     float64
10    float64
11    float64
12    float64
13      int64
14     object
15     object
16      int64
17     object
18    float64
19    float64
20    float64
21    float64
22    float64
23      int64
24      int64
25    float64
dtype: object

In [6]:
df[5] = df[5].replace('?', np.NaN)
df_drop = df.dropna()
df_drop.shape[0]

159

#### Mengisi data missing value di kolom yang bertipe int64 dan float64 dengan nilai mean di masing-masing kolom, sedangkan untuk data di kolom yang bertipe object dengan mode atau modus di masing-masing kolom.

In [7]:
for i in range(len(df.columns)):
    if (df[df[i].isnull()].shape[0]>0):
        print('\nAttribute-',i,' (before) :',df[df[i].isnull()].shape[0])
        if (df[i].dtypes in ['int64','float64']):
            df[i].fillna(df[df[i].notnull()][i].mean(), inplace=True)
        else:
            df[i].fillna(df[df[i].notnull()][i].mode()[0], inplace=True)
        print('\nAttribute-',i,' (after) :',df[df[i].isnull()].shape[0])
    
    df.describe()


Attribute- 1  (before) : 41

Attribute- 1  (after) : 0

Attribute- 5  (before) : 2

Attribute- 5  (after) : 0

Attribute- 18  (before) : 4

Attribute- 18  (after) : 0

Attribute- 19  (before) : 4

Attribute- 19  (after) : 0

Attribute- 21  (before) : 2

Attribute- 21  (after) : 0

Attribute- 22  (before) : 2

Attribute- 22  (after) : 0

Attribute- 25  (before) : 4

Attribute- 25  (after) : 0


#### Mengubah nilai data yang bertipe object menjadi category

In [8]:
for i in range(len(df.columns)):
    if (df[df.columns[i]].dtypes=='object'):
        df[df.columns[i]] = df[df.columns[i]].astype('category')
        df[df.columns[i]] = df[df.columns[i]].cat.codes
df.head()

0      1   2   3   4   5   6   7   8     9   ...   16  17    18    19  \
0   3  122.0   0   1   0   1   0   2   0  88.6  ...  130   5  3.47  2.68   
1   3  122.0   0   1   0   1   0   2   0  88.6  ...  130   5  3.47  2.68   
2   1  122.0   0   1   0   1   2   2   0  94.5  ...  152   5  2.68  3.47   
3   2  164.0   1   1   0   0   3   1   0  99.8  ...  109   5  3.19  3.40   
4   2  164.0   1   1   0   0   3   0   0  99.4  ...  136   5  3.19  3.40   

     20     21      22  23  24       25  
0   9.0  111.0  5000.0  21  27  13495.0  
1   9.0  111.0  5000.0  21  27  16500.0  
2   9.0  154.0  5000.0  19  26  16500.0  
3  10.0  102.0  5500.0  24  30  13950.0  
4   8.0  115.0  5500.0  18  22  17450.0  

[5 rows x 26 columns]

### Prepare the experiments
#### Tujuan Penelitian
* Membandingkan model algoritma Linear Regression dengan algoritma Decision Tree Regressor untuk dataset automobile.
#### Respon / Ukuran Kinerja
* Persentase prediksi harga mobil.
#### Faktor dan Level
* Faktor yang mempengaruhi untuk penerapan algoritma Decision Tree Regressor adalah nilai cv.
* Nilai cv untuk ekksperimen ini ditentukan sebesar 4.
#### Desain Eksperimen
* Skema eksperimen yang digunakan adalah Cross validation.

### Membagi dataset menjadi data training dan data testing

In [9]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

source = df.iloc[:,:-1] 
target = df[25] 

sv_train, sv_test, tv_train, tv_test = train_test_split(source, target, random_state=3, test_size=0.2)
print("\nData Training:\n")
print(sv_train.head())
print(sv_train.shape)

print("\nData Testing:\n")
print(sv_test.head())
print(sv_test.shape)


Data Training:

     0      1   2   3   4   5   6   7   8     9   ...  15   16  17    18  \
135   2  104.0  17   1   0   0   3   1   0  99.1  ...   2  121   5  3.54   
134   3  150.0  17   1   0   1   2   1   0  99.1  ...   2  121   5  2.54   
41    0   85.0   5   1   0   0   3   1   0  96.5  ...   2  110   5  3.15   
161   0   91.0  19   1   0   0   2   1   0  95.7  ...   2   98   1  3.19   
159   0   91.0  19   0   0   0   2   1   0  95.7  ...   2  110   3  3.27   

       19    20     21      22  23  24  
135  3.07   9.3  110.0  5250.0  21  28  
134  2.07   9.3  110.0  5250.0  21  28  
41   3.58   9.0  101.0  5800.0  24  28  
161  3.03   9.0   70.0  4800.0  28  34  
159  3.35  22.5   56.0  4500.0  38  47  

[5 rows x 25 columns]
(164, 25)

Data Testing:

     0      1   2   3   4   5   6   7   8     9   ...  15   16  17    18  \
189   3  122.0  20   1   0   1   0   1   0  94.5  ...   2  109   5  3.19   
183   2  122.0  20   1   0   1   3   1   0  97.3  ...   2  109   5  3.19   
118

### Modeling : Algoritma LinearRegression

In [10]:
lr_model = LinearRegression()
lr_model.fit(sv_train, tv_train)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=None, normalize=False)

In [11]:
lr_model.coef_

array([ 1.89029998e+01, -8.48474427e+00, -2.15483383e+02, -1.45707760e+03,
        2.47086702e+03, -7.30625935e+02, -1.21071269e+03,  5.82372687e+02,
        1.17479259e+04,  8.06324351e+01,  9.94292188e+01,  1.67609530e+02,
        4.05893454e+02,  1.58811792e+00,  5.28804321e+02,  1.16942291e+03,
        1.40550435e+02,  3.08117664e+02, -1.52402495e+03, -3.06856653e+03,
       -1.11672651e+02, -2.85401695e+01,  2.86497480e+00,  8.09874892e+01,
        2.30481979e+01])

In [12]:
lr_model.intercept_

-63148.01127701166

In [13]:
pred = lr_model.predict(sv_test)

In [14]:
pred 

array([10667.93058809,  8391.42335456,  5525.55629663,  7464.77647817,
        6120.10470756, 26114.70117215,  8794.30342131,  5764.85656479,
        6430.75393948, 32546.15745336,  5702.43831079,  7419.07706473,
        5300.48252267, 22542.94980325, 31016.94556977,  9553.52548174,
        5221.07728172, 11185.57922867,  5002.47797605,  8336.60153377,
       24688.89472305, 23204.09278234,  7726.6109416 , 13085.24640407,
       11449.89105046, 14837.56659566, 22056.91324617, 13162.22488402,
       18030.90067162, 22570.51771766,  6805.93336847, 19491.30110177,
       13219.98825805, 41271.494495  , 12093.09297226,  7078.1901166 ,
       11673.14575067, 29002.22824727, 15085.0074932 , 10186.03786959,
        6745.68582203])

#### Menghitung akurasi menggunakan cross validation score R2 untuk model LinearRegression

In [16]:
from sklearn import model_selection

results = model_selection.cross_val_score(lr_model, source, target, cv=4, scoring='r2')
print(results.mean())

0.5475930817607665


### Re-modeling : Algoritma DecisionTreeRegressor

In [17]:
from sklearn.tree import DecisionTreeRegressor

regressor = DecisionTreeRegressor(random_state = 0)
regressor.fit(sv_train, tv_train)

DecisionTreeRegressor(criterion='mse', max_depth=None, max_features=None,
                      max_leaf_nodes=None, min_impurity_decrease=0.0,
                      min_impurity_split=None, min_samples_leaf=1,
                      min_samples_split=2, min_weight_fraction_leaf=0.0,
                      presort=False, random_state=0, splitter='best')

In [18]:
pred = regressor.predict(sv_test)

In [19]:
pred

array([ 9549.        ,  7295.        ,  5389.        ,  9538.        ,
        7198.        , 25552.        ,  8845.        ,  6918.        ,
        7609.        , 36880.        ,  6669.        ,  7603.        ,
        8249.        , 18950.        , 35550.        ,  8495.        ,
        7198.        ,  7895.        ,  8358.        ,  7775.        ,
       19045.        , 18420.        ,  6529.        , 13207.12935323,
        8845.        , 11900.        , 18150.        , 11694.        ,
       21105.        , 18399.        ,  6795.        , 19045.        ,
       16503.        , 45400.        , 13645.        ,  7995.        ,
       11395.        , 36000.        ,  8778.        , 18344.        ,
        7898.        ])

#### Menghitung akurasi menggunakan cross validation score R2 untuk model DecisionTreeRegressor

In [20]:
results = model_selection.cross_val_score(regressor, source, target, cv=4, scoring='r2')
print(results.mean())

0.6282904715924365


### Kesimpulan
* Untuk data automobile ini, modelling pertama menggunakan algoritma Linear Regression dan menghasilkan nilai akurasi menggunakan metode cross validation R2 sebesar 0.547 (55%).
* Setelah didapatkan akurasi sebesar 55% untuk modelling menggunakan Linear Regression, dilakukan Remodeling menggunakan algoritma Decision Tree Regressor dan menghasilkan nilai akurasi yang meningkat, yaitu sebesar 0.628 (63%).
* Jadi untuk data automobile, algoritma Decision Tree Regressor menghasilkan model yang lebih baik dibandingkan dengan menggunakan algoritma Linear Regression.